packages/shared/src/pipelines.ts

```diff
@@ -0,0 +1,95 @@
+export interface Operation<T, D = unknown> {
+  invoke(data: D): T;
+}
+
+export class Pipeline<T, D> implements Operation<T[]> {
+  private operations: Operation<T>[] = [];
+
+  addOperation(operation: Operation<T>) {
+    this.operations.push(operation);
+    return this;
+  }
+
+  invoke(data: D): T[] {
+    return this.operations.map(operation => operation.invoke(data));
+  }
+}
```

Comment:
I'm trying to understand the pattern used here but I do need to do some more reading. This, however, looks like we're simply chaining methods while keeping a reference to their return value. Is this simply it?

Reply:
The Pipeline pattern is a mix of 2 more commonly known patterns:
- [Chain of Responsibility](https://refactoring.guru/design-patterns/chain-of-responsibility)
- [Command](https://refactoring.guru/design-patterns/command)

>This, however, looks like we're simply chaining methods while keeping a reference to their return value.
(kind of) yes.

Article about Pipeline pattern: https://medium.com/@bonnotguillaume/software-architecture-the-pipeline-design-pattern-from-zero-to-hero-b5c43d8a4e60